lib/Parser/BGParseManager.h

```diff
@@ -0,0 +1,133 @@
+// multi-threaded; please see the definition of each function for expectations of which thread executes
+// the function.
+//
+// There are up to 3 threads involved per background parse. Here are the relevant BGParseManager functions
+// called from these three threads:
+//
+//      Background/Network          JobProcessor                UI/Executing 
+//      Thread                      Thread                      Thread
+//              |                       |                           |
+//      QueueBackgroundParse            |                           |
+//              |                       |                           |
+//              |                   Process                         |
+//              |                       |                           |
+//              |                       |                       GetParseResults
+//              .                       .                           .
+// Note that the thread queueing the work can also be the UIT thread. Also, note that GetParseResults may
+// block the calling thread until the JobProcessor thread finishes processing the BGParseWorkItem that
+// contains the results.
+
+
+// Forward Declarations
+class BGParseWorkItem;
+namespace Js
+{
+    struct Utf8SourceInfo;
+    typedef void * Var;
+    class JavascriptFunction;
+}
+
+
+// BGParseManager is the primary interface for background parsing. It uses a cookie to publicly track the data
+// involved per parse request.
+class BGParseManager sealed : public JsUtil::WaitableJobManager
+{
+public:
+    BGParseManager();
+    ~BGParseManager();
+
+    static BGParseManager* GetBGParseManager();
+    static void DeleteBGParseManager();
+    static DWORD GetNextCookie();
+
+    HRESULT QueueBackgroundParse(LPCUTF8 pszSrc, size_t cbLength, char16 *fullPath, DWORD* dwBgParseCookie);
+    HRESULT GetInputFromCookie(DWORD cookie, LPCUTF8* ppszSrc, size_t* pcbLength);
+    HRESULT GetParseResults(
+        Js::ScriptContext* scriptContextUI,
+        DWORD cookie,
+        LPCUTF8 pszSrc,
+        SRCINFO const * pSrcInfo,
+        Js::ParseableFunctionInfo** ppFunc,
+        CompileScriptException* pse,
+        size_t& srcLength
+    );
+
+    virtual bool Process(JsUtil::Job *const job, JsUtil::ParallelThreadData *threadData) override;
+    virtual void JobProcessed(JsUtil::Job *const job, const bool succeeded) override;
+    virtual void JobProcessing(JsUtil::Job* job) override;
+
+    // Defines needed for jobs.inl
+    BGParseWorkItem* GetJob(BGParseWorkItem* workitem);
+    bool WasAddedToJobProcessor(JsUtil::Job *const job) const;
+
+private:
+    BGParseWorkItem * FindJob(DWORD dwCookie, bool waitForResults);
+
+    // BGParseWorkItem job can be in one of 3 states, based on which linked list it is in:
+    // - queued - JobProcessor::jobs
+    // - processing - BGParseManager::workitemsProcessing
+    // - processed - BGParseManager::workitemsProcessed
+    JsUtil::DoublyLinkedList<JsUtil::Job> workitemsProcessing;
```

Comment:
Could we make these lists be of BGParseWorkItem to reduce the number of places we need to cast? (I also wouldn't mind if base classes like JobProcessor were templated on the job type, but I imagine that change might get obnoxiously big.)

Reply:
no objections